README.md
## oncoEnrichR - cancer-dedicated gene set interpretation

### Contents

-   [Overview](#overview)
-   [News](#news)
-   [Annotation resources](#annotation-resources)
-   [Example report](#example-report)
-   [Getting started](#getting-started)
-   [Citation](#citation)
-   [Contact](#contact)
-   [Funding and Collaboration](#funding-and-collaboration)
-   [Code of Conduct](#code-of-conduct)

### Overview

**oncoEnrichR** is an R package for functional interrogation of human genesets in the context of cancer. The package leverages an extensive amount of prior molecular knowledge for geneset interpretation along multiple dimensions.

Web-based access to **oncoEnrichR** is available at <https://oncotools.elixir.no>

**oncoEnrichR** is primarily intended for exploratory analysis and prioritization of a gene list (referred to here as **query set**) from high-throughput cancer biology experiments, e.g. genetic screens (siRNA/CRISPR), protein proximity labeling, or transcriptomics (differential expression). The tool queries a number of high-quality data resources in order to assemble useful gene annotations and analyses in an interactive report (examples from the report shown below).

<br>

<img src="man/figures/oncoenrichr_overview.png" align="center" border="1,"/>

<br><br>

The contents of the final report attempt to provide answers to the following questions:

-   Which diseases/tumor types are known to be associated with genes in the query set, and to what extent?
-   Which genes in the query set are attributed with cancer hallmark evidence?
-   Which proteins in the query sets are druggable in diffferent cancer conditions (early and late clinical development phases)? For other proteins in the query set, what is their likelihood of being druggable?
-   Which protein complexes involve proteins in the query set?
-   Which subcellular compartments (nucleus, cytosol, plasma membrane etc) are dominant localizations for proteins in the query set?
-   Are specific tissues or cell types enriched in the query set, considering tissue/cell-type specific expression patterns of target genes?
-   Which protein-protein interactions are known within the query set? Are there interactions between members of the query set and other cancer-relevant proteins (e.g. proto-oncogenes, tumor-suppressors or predicted cancer drivers)? Which proteins constitute hubs in the protein-protein interaction network?
-   Which known regulatory interactions (TF-target) are found within the query set, and what is their mode of regulation (repressive vs. stimulating)?
-   Are there occurrences of known ligand-receptor interactions within the query set?
-   Are there specific pathways, biological processes, or pre-defined molecular signatures that are enriched within the query set, as compared to a reference/background set?
-   Which members of the query set are frequently mutated in tumor sample cohorts (TCGA, SNVs/InDels, homozygous deletions, copy number amplifications)?
-   Which members of the query set are co-expressed (strong negative or positive correlations) with cancer-relevant genes (i.e. proto-oncogenes or tumor suppressors) in tumor sample cohorts (TCGA)?
-   Which members of the query set are associated with better/worse survival in different cancers, considering high or low gene expression levels, mutation, or copy number status in tumors?
-   Which members of the query set are associated with cellular loss-of-fitness in CRISPR/Cas9 whole-genome drop out screens of cancer cell lines (i.e. reduction of cell viability elicited by a gene inactivation)? Which targets are prioritized therapeutic targets, considering fitness effects and genomic biomarkers in combination?

### News

-   November 30th 2021: **1.0.7 release**
    
    -   Data updates: Open Targets Platform, UniProt KB, WikiPathways, CancerMine, TCGA, EFO, Human Protein Atlas
    -   Re-design of data management in the package, analysis now requires an initial *database loading step* 

-   October 27th 2021: **1.0.6 release**

    -   Addition of new modules
        -   Regulatory interactions - DoRothEA
        -   Ligand-receptor interactions - CellChatDB
    -   Data updates: Open Targets Platform, KEGG, WikiPathways, CancerMine, TCGA, EFO, DiseaseOntology
    -   Additional dedicated enrichment plots (GO)
    -   Improved target rankings (tissue- and celltype enrichment, regulatory interactions)
    -   See complete [CHANGELOG](https://sigven.github.io/oncoEnrichR/news/index.html)

-   July 8th 2021: **1.0.4 release**

    -   Fixed a bug in v1.0.3 - HTML report generation failed

-   June 29th 2021: **1.0.3 release**

    -   Minor typo fixes in HTML output
    -   Data upgrade: KEGG, WikiPathways, CancerMine
    -   **NEW**: web interface to oncoEnrichR: [oncotools.elixir.no](https://oncotools.elixir.no)

-   June 8th 2021: **1.0.0 release**

    -   Added methylation as prognostic feature from [Smith et al., *bioRxiv*, 2021](https://www.biorxiv.org/content/10.1101/2021.06.01.446243v1)
    -   To keep the size of output Excel/HTML files at managable levels, maximum size of queryset is reduced to **n = 600 genes**
    -   Fixed assymetrical scales of prognostic Z-score plots

### Annotation resources

Data harvested from the following resources form the backbone of *oncoEnrichR*:

-   [Open Targets Platform](https://targetvalidation.org) - human drug-target associations and comprehensive disease-target associations
-   [CancerMine](http://bionlp.bcgsc.ca/cancermine/) - literature-mined database of drivers, oncogenes and tumor suppressors in cancer
-   [DoRothEA](https://saezlab.github.io/dorothea/index.html) - gene set resource containing signed transcription factor (TF) - target (regulatory) interactions
-   [CellChatDB](http://www.cellchat.org/) - ligand-receptor interaction resource
-   [The Cancer Genome Atlas (TCGA)](https://portal.gdc.cancer.gov/) - gene aberration frequencies and co-expression patterns in \~9,500 primary tumor samples
-   [The Human Protein Atlas](https://www.proteinatlas.org) - expression data for healthy human tissues ([GTex](https://gtexportal.org/home/))/cell types, and prognostic gene expression associations in cancer
-   [Molecular Signatures Database (MSigDB)](http://software.broadinstitute.org/gsea/msigdb/index.jsp) - collection of annotated (e.g. towards pathways) genesets for enrichment/overrepresentation analysis. This includes genesets from [Gene Ontology](http://geneontology.org/), [Reactome](https://reactome.org/), [KEGG](https://www.genome.jp/kegg/pathway.html), [WikiPathways](https://www.wikipathways.org/index.php/WikiPathways), [BIOCARTA](https://maayanlab.cloud/Harmonizome/dataset/Biocarta+Pathways), as well as curated [immunologic](https://www.gsea-msigdb.org/gsea/msigdb/collections.jsp#C7) and [cancer-specific](https://www.gsea-msigdb.org/gsea/msigdb/collections.jsp#C6) signatures.
-   [NetPath](http://www.netpath.org) - signaling transduction pathways
-   [STRING](https://string-db.org) - protein-protein interaction database
-   [CORUM](https://mips.helmholtz-muenchen.de/corum/) - protein complex database
-   [Compleat](https://fgr.hms.harvard.edu/compleat%3E) - protein complex resource
-   [ComplexPortal](https://www.ebi.ac.uk/complexportal/home) - manually curated, encyclopaedic resource of macromolecular complexes
-   [hu.MAP2](http://humap2.proteincomplexes.org) - human protein complex map
-   [ComPPI](http://comppi.linkgroup.hu/) - subcellular compartment database
-   [Project Score](https://score.depmap.sanger.ac.uk) - Database on the effects on cancer cell line viability elicited by CRISPR-Cas9 mediated gene activation
-   [Genetic determinants of survival in cancer](http://survival.cshl.edu/) - Resource on the relative prognostic impact of gene mutation, expression, methylation, and CNA in human cancers

### Example report

<a href="https://doi.org/10.5281/zenodo.5737599"><img src="https://zenodo.org/badge/DOI/10.5281/zenodo.5737599.svg" alt="DOI"/></a>

<br>

<img src="man/figures/oncoenrichr_slideshow2.gif" align="center" width="560" height="750" border="2,"/>

### Getting started

#### Installation

##### Latest release

1.  `install.packages('devtools')`
2.  `devtools::install_github('sigven/oncoEnrichR', ref = "v1.0.7")`
3.  `library(oncoEnrichR)`

[![stable](http://badges.github.io/stability-badges/dist/stable.svg)](http://github.com/badges/stability-badges)

#### Usage

*oncoEnrichR* performs its operations through the following procedures/methods:

**0.** `oncoEnrichR::load_db()`

-   Loads (from web) all the necessary annotation datasets used in oncoEnrichR

**1.** `oncoEnrichR::onco_enrich()`

-   Consists of two main processing steps:

    1\) Takes an input/query list of human gene/protein identifiers (e.g. UniProt accession, RefSeq/Ensembl transcript identifer etc.) as input and conducts uniform identifier conversion

    2\) Performs extensive annotation, enrichment and membership analyses of the query set against underlying data sources on cancer-relevant properties of human genes and their interrelationships.

-   Technically, the method returns a *list object* with all contents of the analyses performed. The specific arguments/options and default values are outlined below:

    ``` r
    onco_enrich(
      query,
      oeDB,
      query_id_type = "symbol",
      ignore_id_err = TRUE,
      html_floating_toc = TRUE,
      html_report_theme = "default",
      project_title = "Project title",
      project_owner = "Project owner",
      project_description = "Project description",
      bgset = NULL,
      bgset_id_type = "symbol",
      bgset_description = "All protein-coding genes",
      p_value_cutoff_enrichment = 0.05,
      p_value_adjustment_method = "BH",
      q_value_cutoff_enrichment = 0.2,
      min_geneset_size = 10,
      max_geneset_size = 500,
      num_terms_enrichment_plot = 20,
      min_subcellcomp_confidence = 1,
      subcellcomp_show_cytosol = FALSE,
      min_confidence_reg_interaction = "D",
      simplify_go = F,
      ppi_add_nodes = 50,
      ppi_score_threshold = 900,
      show_ppi = T,
      show_drugs_in_ppi = T,
      show_disease = T,
      show_top_diseases_only = T,
      show_cancer_hallmarks = T,
      show_drug = T,
      show_enrichment = T,
      show_tcga_aberration = T,
      show_tcga_coexpression = T,
      show_cell_tissue = F,
      show_ligand_receptor = T,
      show_regulatory_interactions = T,
      show_unknown_function = T,
      show_prognostic_cancer_assoc = T,
      show_subcell_comp = T,
      show_crispr_lof = T,
      show_complex = T)
    ```

    See [detailed descriptions of all options here](https://sigven.github.io/oncoEnrichR/reference/onco_enrich.html)

**2.** `oncoEnrichR::write()`

-   Consists of two main processing steps:

    1\) Transform the contents of the analyses returned by *oncoEnrichR::onco_enrich()* into various visualizations and interactive tables

    2\) Assemble and write the final analysis report through

    -   A\) a structured and interactive *oncoEnrichR* HTML report, OR
    -   B\) a multisheet Excel workbook with results from the annotations and analyses provided by *oncoEnrichR*

#### Example run

A target list of *n = 134* high-confidence interacting proteins with the c-MYC oncoprotein were previously identified through BioID protein proximity assay in standard cell culture and in tumor xenografts ([Dingar et al., J Proteomics, 2015](https://www.ncbi.nlm.nih.gov/pubmed/25452129)). We ran this target list through the *oncoEnrichR* analysis workflow using the following configurations for the `onco_enrich` method:

-   `project_title = "cMYC_BioID_screen"`
-   `project_owner = "Raught et al."`

and produced the [following HTML report with results](https://doi.org/10.5281/zenodo.5737599).

Below are R commands provided to reproduce the example output. **NOTE**: Replace "LOCAL_FOLDER" with a directory on your local computer:

-   `library(oncoEnrichR)`
-   `myc_interact_targets <- read.csv(system.file("extdata","myc_data.csv", package = "oncoEnrichR"), stringsAsFactors = F)`
-   `oeDB <- oncoEnrichR::load_db()` **NOTE**: pass a writable directory to the argument `cache_dir` in order to speed the database loading in subsequent R sessions
-   `myc_report <- oncoEnrichR::onco_enrich(query = myc_interact_targets$symbol, oeDB = oeDB, show_cell_tissue = T, project_title = "cMYC_BioID_screen", project_owner = "Raught et al.")`
-   `oncoEnrichR::write(report = myc_report, file = "LOCAL_FOLDER/myc_report_oncoenrichr.html", format = "html")`
-   `oncoEnrichR::write(report = myc_report, file = "LOCAL_FOLDER/myc_report_oncoenrichr.xlsx", format = "excel")`

### Citation

If you use oncoEnrichR, please cite the following preprint:

<https://arxiv.org/abs/2107.13247>

### Contact

sigven AT ifi.uio.no

### Funding and Collaboration

OncoEnrichR is supported by the [Centre for Cancer Cell Reprogramming](https://www.med.uio.no/cancell/english/) at the [University of Oslo](https://www.uio.no)/[Oslo University Hospital](https://radium.no), and [Elixir Norway (Oslo node)](https://elixir.no/organization/organisation/elixir-uio).

<br> <br>

<p float="left">
  <a href="https://www.med.uio.no/cancell/english/">
     <img src="man/figures/can-cell.png" width="150" >
  </a>
  &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
  <a href="https://elixir.no/organization/organisation/elixir-uio">
     <img src="man/figures/elixir_norway.png" width="200" />
  </a>
</p>

### Code of Conduct

Please note that this project is released with a [Contributor Code of Conduct](https://github.com/sigven/oncoEnrichR/blob/master/.github/CODE_OF_CONDUCT.md). By participating in this project you agree to abide by its terms.


NEWS.md

# Version 1.0.7

## Added

* Data updates
  * WikiPathways (20211110)
  * Open Targets Platform (2021.11)
  * EFO (v3.36.0)
  * TCGA (v31, 2021-10-29)
  * CancerMine (v40)
  * Human Protein Atlas (v21)
  * UniProt KB (2021_04)
  * PFAM domains (release 35, November 2021)
  
## Changed

- Management of annotation databases have been re-designed, reducing the size 
and installation of *oncoEnrichR* significantly. The re-design results in
the addition of an initial step in the analysis; **database loading**.


# Version 1.0.6

## Added

* Data updates 
  * KEGG (20211013)
  * WikiPathways (20211010)
  * Open Targets Platform (2021.09)
  * EFO (v3.35.0), DiseaseOntology (v2021-10-11)
  * TCGA (v30, 2021-09-23)
  * Protein complex annotations - additions from [Compleat](https://fgr.hms.harvard.edu/compleat), [ComplexPortal](https://www.ebi.ac.uk/complexportal/home), [PDB](https://www.rcsb.org/), [hu.MAP2](http://humap2.proteincomplexes.org/)

* New functionality
  * Report modules
     * **Regulatory interactions** - ([DoRothEA](https://saezlab.github.io/dorothea/))
        * View overlap between members of queryset with previously established regulatory interactions (TF-target relationships) from the DoRothEA resource
     * **Ligand-receptor interactions** -  ([CellChatDB](http://www.cellchat.org/))
        * Explore potential ligand-receptor interactions in the query set, as found in the CellChat database

  * Report styling options
     * Choose your own Bootswatch theme for HTML report - option `html_report_theme`
     * Choose where the table of contents are placed in the HTML report (floating-left or static at the top) - option `html_floating_toc`

  * Arguments to `oncoEnichR::onco_enrich()`:
     * `html_floating_toc` - logical, if set to FALSE, table of contents in HTML report will be placed on top of the main document
     * `html_report_theme` - character, choose between different Bootswatch themes for style in HTML report
     * `show_regulatory` - logical, show regulatory interactions module (DoRothEA)
     * `min_confidence_reg_interaction` - character, minimum confidence of regulatory interactions included from DoRothEA ('A','B','C', or 'D')
     * `show_ligand_receptor` - logical, show ligand-receptor interaction module (CellChatDB)
     * `num_terms_enrichment_plot` - integer, number of enriched Gene Ontology terms (max) to show in enrichment barplots (module *Functional Enrichment*)

  * Default gene rankings
     * In the modules **Regulatory Interactions** and **Tissue and cell type enrichment**, interactions/targets are ranked according to a quantitative cancer association score *cancer_max_rank* (maximum gene-cancer association rank (Open Targets Platform) across primary sites)

## Fixed
  * Report *disclaimer* occurring at the bottom of the report is no longer missing from web-based analysis (oncotools.elixir.no) - file `_site.yml` in `inst/templates`

# Version 1.0.5

## Added

* Data updates: KEGG (20210916), WikiPathways (20210910), CancerMine (v39),
EFO (v3.34.0)

# Version 1.0.4

## Added

* Data updates: Open Targets Platform (21.06), CancerMine (v37)
* Software upgrades: multiple R packages, including clusterProfiler (4.0.2)

# Version 1.0.3

## Added

* Functionality for Galaxy-specific HTML output (non-selfcontained HTML)
* Data updates: WikiPathways (20210610), KEGG (20210610), CancerMine (v36, 20210611)
* Web-based interface to oncoEnrichR available: [oncotools.elixir.no](https://oncotools.elixir.no)

## Fixed

* Minor typos in HTML output
* Ambiguous Entrez gene identifiers (towards gene symbols)

# Version 1.0.0

## New

* Added prognostic Z-scores also for methylation features ([Smith et al., _bioRxiv_, 2021](https://www.biorxiv.org/content/10.1101/2021.06.01.446243v1))
* To keep the file size of output Excel/HTML at managable levels, maximum size of queryset is reduced to **n = 600 genes**

## Fixed

* Fixed assymetrical scales of prognostic Z-score plots
* To decrease the size of the HTML report, limited datatable output of recurrent somatic SNVs/InDels and sCNA frequencies to top 2,500 (full set is kept in Excel output)

# Version 0.9.7 (June 4th 2021) 

## New

* More prognostic associations - harvested from [Smith et al., _bioRxiv_, 2021](https://www.biorxiv.org/content/10.1101/2021.06.01.446243v1)

## Fixed

* DESCRIPTION: Added `TissueEnrich` and `assertable` dependency

## Data upgrade

* Target-disease associations from Open Targets Platform (v2021.04)
* WikiPathways (20210510)
* Disease Ontology (20210603)
* KEGG (20210531)



R/tf_targets.R

annotate_tf_targets <- function(qgenes,
                                genedb = NULL,
                                tf_target_interactions = NULL,
                                collection = "global",
                                min_confidence_reg_interaction = "D",
                                logger = NULL){

  stopifnot(is.character(collection))
  stopifnot(collection == "global" | collection == "pancancer")

  stopifnot(is.character(qgenes))
  stopifnot(!is.null(genedb))
  stopifnot(!is.null(logger))
  stopifnot(!is.null(tf_target_interactions[[collection]]))
  validate_db_df(genedb, dbtype = "genedb")
  validate_db_df(tf_target_interactions[[collection]], dbtype = "dorothea")

  log4r_info(logger,
    paste0("DoRothEA: retrieval of regulatory interactions involving members of target set - ", collection))

  exclude_level_regex <- "E"
  if(min_confidence_reg_interaction == "C"){
    exclude_level_regex <- "D|E"
  }
  if(min_confidence_reg_interaction == "B"){
    exclude_level_regex <- "C|D|E"
  }
  if(min_confidence_reg_interaction == "A"){
    exclude_level_regex <- "B|C|D|E"
  }

  target_genes <- data.frame("target" = qgenes, stringsAsFactors = F) %>%
    dplyr::left_join(tf_target_interactions[[collection]],
                     by = c("target")) %>%
    dplyr::distinct() %>%
    dplyr::filter(!is.na(.data$confidence_level)) %>%
    dplyr::filter(!stringr::str_detect(.data$confidence_level, exclude_level_regex))

  if(nrow(target_genes) > 0){
    target_genes <- target_genes %>%
      dplyr::mutate(queryset_overlap = paste("TARGET", .data$confidence_level, sep = "_"))
  }

  tf_genes <- data.frame("regulator" = qgenes, stringsAsFactors = F) %>%
    dplyr::left_join(tf_target_interactions[[collection]],
                     by = c("regulator")) %>%
    dplyr::distinct() %>%
    dplyr::filter(!stringr::str_detect(.data$confidence_level, exclude_level_regex))

  if(nrow(tf_genes) > 0){
    tf_genes <- tf_genes %>%
      dplyr::mutate(queryset_overlap = paste("TF", .data$confidence_level, sep = "_"))
  }

  query_tf_target_interactions <- data.frame()

  if(nrow(target_genes) > 0 | nrow(tf_genes) > 0){
    query_tf_target_interactions <- as.data.frame(
      tf_genes %>%
        dplyr::bind_rows(target_genes) %>%
        dplyr::group_by(.data$regulator, .data$target, .data$mode_of_regulation,
                        .data$interaction_sources, .data$confidence_level,
                        .data$tf_target_literature_support) %>%
        dplyr::summarise(queryset_overlap = paste(.data$queryset_overlap, collapse=";"),
                         .groups = "drop") %>%
        dplyr::left_join(dplyr::select(genedb,
                                       .data$symbol, .data$genename,
                                       .data$cancer_max_rank),
                         by = c("regulator" = "symbol")) %>%
        dplyr::rename(regulator_name = .data$genename,
                      regulator_cancer_max_rank = .data$cancer_max_rank) %>%
        dplyr::left_join(dplyr::select(genedb,
                                       .data$symbol, .data$genename,
                                       .data$cancer_max_rank),
                         by = c("target" = "symbol")) %>%
        dplyr::rename(target_name = .data$genename,
                      target_cancer_max_rank = .data$cancer_max_rank) %>%
        dplyr::mutate(queryset_overlap = stringr::str_replace(
          .data$queryset_overlap, "(A|B|C|D);","")) %>%
        dplyr::mutate(queryset_overlap = factor(
          .data$queryset_overlap, levels = c("TF_TARGET_A","TF_TARGET_B","TF_TARGET_C",
                                       "TF_TARGET_D","TF_A","TF_B","TF_C","TF_D",
                                       "TARGET_A","TARGET_B","TARGET_C","TARGET_D"))) %>%
        dplyr::arrange(.data$queryset_overlap, .data$confidence_level,
                       dplyr::desc(.data$regulator_cancer_max_rank),
                       dplyr::desc(.data$target_cancer_max_rank)) %>%
        dplyr::rename(literature_support = .data$tf_target_literature_support) %>%
        dplyr::select(.data$regulator, .data$regulator_name,
                      .data$target, .data$target_name,
                      .data$confidence_level, .data$mode_of_regulation,
                      .data$literature_support, .data$interaction_sources,
                      .data$queryset_overlap, .data$regulator_cancer_max_rank,
                      .data$target_cancer_max_rank)
    )
  }


  return(query_tf_target_interactions)

}

retrieve_tf_target_network <- function(tf_target_interactions = NULL){


  tf_target_network <- list()
  tf_target_network[['nodes']] <- data.frame()
  tf_target_network[['edges']] <- data.frame()

  if(NROW(tf_target_interactions) > 0){

    complete_interactions <- tf_target_interactions %>%
      dplyr::filter(stringr::str_detect(.data$queryset_overlap,"TF_TARGET_")) %>%
      utils::head(200)

    if(NROW(complete_interactions) > 0){

      tf_target_network[['edges']] <- complete_interactions %>%
        dplyr::select(.data$regulator, .data$target,
                      .data$confidence_level, .data$mode_of_regulation) %>%
        dplyr::rename(from = .data$regulator,
                      to = .data$target) %>%
        dplyr::mutate(title = paste0(.data$mode_of_regulation, ": confidence level ",
                                     .data$confidence_level)) %>%
        dplyr::mutate(arrows = "to") %>%
        dplyr::mutate(length = dplyr::case_when(
          .data$confidence_level == "A" ~ 150,
          .data$confidence_level == "B" ~ 200,
          .data$confidence_level == "C" ~ 250,
          .data$confidence_level == "D" ~ 300,
        )) %>%
        dplyr::mutate(dashes = T) %>%
        dplyr::mutate(color = dplyr::case_when(
          .data$mode_of_regulation == "Stimulation" ~ "darkgreen",
          .data$mode_of_regulation == "Repression" ~ "darkred",
         )) %>%
        dplyr::arrange(.data$confidence_level) %>%
        utils::head(150)

      all_nodes <-
      #tf_target_network[['nodes']] <-
        data.frame('id' = complete_interactions$regulator,
                   'shadow' = F,
                   'size' = 25,
                   'label' = complete_interactions$regulator,
                   'title' = stringr::str_trim(
                     textclean::replace_html(complete_interactions$regulator_name)
                   ),
                   stringsAsFactors = F) %>%
      dplyr::bind_rows(
        data.frame('id' = complete_interactions$target,
                   'shadow' = F,
                   'size' = 25,
                   'label' = complete_interactions$target,
                   'title' = stringr::str_trim(
                     textclean::replace_html(complete_interactions$target_name)
                   ),
                   stringsAsFactors = F)) %>%
        dplyr::distinct()

      nodeset1 <- all_nodes %>%
        dplyr::inner_join(
          dplyr::select(tf_target_network[['edges']], .data$from),
          by = c("id" = "from")) %>%
        dplyr::distinct()

      nodeset2 <- all_nodes %>%
        dplyr::inner_join(
          dplyr::select(tf_target_network[['edges']], .data$to),
          by = c("id" = "to")) %>%
        dplyr::distinct()

      tf_target_network[['nodes']] <-
        nodeset1 %>%
        dplyr::bind_rows(nodeset2) %>%
        dplyr::distinct()


      regulators <- tf_target_network[['edges']] %>%
        dplyr::select(.data$from) %>%
        dplyr::inner_join(
          dplyr::select(tf_target_network[['nodes']], .data$id),
          by = c("from" = "id")
          ) %>%
        dplyr::rename(id = .data$from) %>%
        dplyr::distinct()

      targets <- tf_target_network[['edges']] %>%
        dplyr::select(.data$to) %>%
        dplyr::inner_join(
          dplyr::select(tf_target_network[['nodes']], .data$id),
          by = c("to" = "id")
        ) %>%
        dplyr::rename(id = .data$to) %>%
        dplyr::distinct()

      target_and_regulator <- targets %>%
        dplyr::inner_join(regulators, by = "id")

      if(nrow(target_and_regulator) > 0){
        regulators <- regulators %>%
          dplyr::anti_join(target_and_regulator, by = "id") %>%
          dplyr::mutate(group = "Regulator")

        targets <- targets %>%
          dplyr::anti_join(target_and_regulator, by = "id") %>%
          dplyr::mutate(group = "Target")

        target_and_regulator <- target_and_regulator %>%
          dplyr::mutate(group = "Regulator/target")
      }else{
        regulators <- regulators %>%
          dplyr::mutate(group = "Regulator")

        targets <- targets %>%
          dplyr::mutate(group = "Target")
      }

      all_shapes <- regulators %>%
        dplyr::bind_rows(targets) %>%
        dplyr::bind_rows(target_and_regulator)

      tf_target_network[['nodes']] <- tf_target_network[['nodes']] %>%
        dplyr::left_join(all_shapes, by = "id")

    }

  }

  return(tf_target_network)

}
